compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -143,7 +187,35 @@ impl<'f> LoopInvariantContext<'f> {
             is_loop_invariant &=
                 !self.defined_in_loop.contains(&value) || self.loop_invariants.contains(&value);
         });
-        is_loop_invariant && instruction.can_be_deduplicated(&self.inserter.function.dfg, false)
+
+        let can_be_deduplicated = instruction
+            .can_be_deduplicated(&self.inserter.function.dfg, false)
+            || self.can_be_deduplicated_from_upper_bound(&instruction);
+
+        is_loop_invariant && can_be_deduplicated
+    }
+
+    /// Certain instructions can take advantage of that our induction variable has a fixed maximum.
+    ///
+    /// For example, an array access can usually only be safely deduplicated when we have a constant
+    /// index that is below the length of the array.
+    /// Checking an array get where the index is the loop's induction variable on its own
+    /// would determine that the instruction is not safe for hoisting.
+    /// However, if we know that the induction variable's upper bound will always be in bounds of the array
+    /// we can safely hoist the array access.
+    fn can_be_deduplicated_from_upper_bound(&self, instruction: &Instruction) -> bool {
+        match instruction {
+            Instruction::ArrayGet { array, index } => {
+                let array_typ = self.inserter.function.dfg.type_of_value(*array);
+                let upper_bound = self.outer_induction_variables.get(index);
+                if let (Type::Array(_, len), Some(upper_bound)) = (array_typ, upper_bound) {
+                    upper_bound.to_u128() as usize <= len
+                } else {
+                    false
+                }
+            }
+            _ => false,
+        }
     }
 
     fn map_dependent_instructions(&mut self) {
```

Comment:
The `FunctionInserter` maintains a map of old values to new values. As per your name suggestion `resolve_terminator_in_place`, we are resolving the values of the terminator, but we are taking these resolved values to map each previous terminator value to the new one specified by the `FunctionInserter`. I see though how it can look like the method is doing nothing. I have included a comment, let me know if that is more clear.